validation/start.go

```diff
@@ -0,0 +1,110 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+
+	"github.com/mndrix/tap-go"
+	rspecs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/runtime-tools/specerror"
+	"github.com/opencontainers/runtime-tools/validation/util"
+	uuid "github.com/satori/go.uuid"
+)
+
+func wrap(t *tap.T, expected bool, specErr error, detailedErr error) {
+	t.Ok(expected, specErr.(*specerror.Error).Err.Err.Error())
+	diagnostic := map[string]string{
+		"reference": specErr.(*specerror.Error).Err.Reference,
+	}
+
+	if detailedErr != nil {
+		diagnostic["error"] = detailedErr.Error()
+		if e, ok := detailedErr.(*exec.ExitError); ok {
+			if len(e.Stderr) > 0 {
+				diagnostic["stderr"] = string(e.Stderr)
+			}
+		}
+	}
+	t.YAML(diagnostic)
+}
+
+func main() {
+	t := tap.New()
+	t.Header(0)
+
+	bundleDir, err := util.PrepareBundle()
+	if err != nil {
+		util.Fatal(err)
+	}
+	defer os.RemoveAll(bundleDir)
+
+	containerID := uuid.NewV4().String()
+
+	r, err := util.NewRuntime(util.RuntimeCommand, bundleDir)
+	if err != nil {
+		util.Fatal(err)
+	}
+	g := util.GetDefaultGenerator()
+	g.SetProcessArgs([]string{"sh", "-c", fmt.Sprintf("echo 'process called' >> %s", "/output")})
+	r.SetConfig(g)
+	output := filepath.Join(bundleDir, g.Spec().Root.Path, "output")
+
+	// start without id
+	err = r.Start()
+	wrap(t, err != nil, specerror.NewError(specerror.StartWithoutIDGenError, fmt.Errorf("start` operation MUST generate an error if it is not provided the container ID"), rspecs.Version), err)
+
+	// set id for the remaining tests
+	r.SetID(containerID)
+
+	// start a not `created` container - case one: non-exist container
+	err = r.Start()
+	wrap(t, err != nil, specerror.NewError(specerror.StartNotCreatedGenError, fmt.Errorf("attempting to `start` a container that is not `created` MUST generate an error"), rspecs.Version), err)
+
+	r.Create()
+	// start a `created` container
+	err = r.Start()
+	outputData, _ := ioutil.ReadFile(output)
```

Comment:
This is racy.  `start` should [block until the user-specified program has executed][1], but there's no guarantee about how far along that program is in its execution.  That means this read could be happening either before or after the user-specified program gets around to writing to `output`.  The coarse guard for this would be to sleep for some safe-enough time.  The better, but maybe Linux-specific,  guard would be to use `inotify(7)` to block until the file was updated (or some safe-enough time had passed) and compare the measured write time with the `start`-invocation time (the write time should be after `start` was invoked and also after any `prestart` hooks completed).

[1]: https://github.com/opencontainers/runtime-tools/blame/64e1103b1518c53911bd163765ae6391c3d54b19/docs/command-line-interface.md#L159

Reply:
I'll add time sleep here.